pkg/cloud/services/ec2/instances.go

```diff
@@ -358,19 +358,35 @@ func (s *Service) runInstance(role string, i *infrav1.Instance) (*infrav1.Instan
 		}
 	}
 
-	if i.RootDeviceSize != 0 {
+	if i.RootVolumeOptions != nil && i.RootVolumeOptions.Size != 0 {
```

Comment:
so we'll ignore other options if Size==0 ?

Reply:
If the size is 0 we don't create the disk so there's nothing for the options to apply to / be relevant to afaik?

Reply:
so there is no default size? If so, shall we mark size as non-optional?

Reply:
The default size from AWS is actually 8, if that info helps. This 0 check probably originally should have been a nil check with Size as a nil-able field. I forget if there was a reason why I didn't implement it that way the first time.

Reply:
From ec2 sdk:  	
        // The size of the volume, in GiB.
	//
	// Default: If you're creating the volume from a snapshot and don't specify
	// a volume size, the default is the snapshot size.
	//
	// Constraints: 1-16384 for General Purpose SSD (gp2), 4-16384 for Provisioned
	// IOPS SSD (io1), 500-16384 for Throughput Optimized HDD (st1), 500-16384 for
	// Cold HDD (sc1), and 1-1024 for Magnetic (standard) volumes. If you specify
	// a snapshot, the volume size must be equal to or larger than the snapshot
	// size.
	VolumeSize *int64 `locationName:"volumeSize" type:"integer"`

Reply:
Fine however we want to implement this, but if you're always going to get a disk maybe we should make it mandatory so there's not ambiguous behavior with the other options etc?

Reply:
I'm fine having the struct and Size default to nil. Let's make sure to copy the default behavior from the SDK to our Size field.

Reply:
Still not 100% clear on what our behavior should be though. If I specify 0 (currently, or nil after these changes) for the size, I'm going to get a volume with 8 (according to AWS default? Where does this come from? Want to avoid random hard-coded values or copying too much AWS logic). Should we still apply the other options (encryption etc? assuming yes). What's the difference (from a user POV) between size = 0 and size = nil? Does allowing 0 even make sense (what am I expecting if I put 0 for a size?)

Reply:
these are good questions. just my understanding, please correct me if i'm wrong: size=0 and size=nil won't make much difference from user's PoV, both will create a volume with the size of the snapshot that backs the AMI. 

provide option to use default size is valuable, because the non-zero size specified must be >= snapshot size. it's quite common that we don't know the underline AMI snapshot size, so use default size is a safe option.

Reply:
Actually, I'll defer to @detiber and @randomvariable for this, the most important part for me is to be consistent with the experience we're providing to our users